In [6]:
# import h2o package and specific estimator 
import h2o
from h2o.automl import H2OAutoML
import random, os, sys
from datetime import datetime
import pandas as pd
import logging
import csv
import optparse
import time
import json
from distutils.util import strtobool
import psutil

import warnings
warnings.filterwarnings('ignore')

In [5]:
h2o.init(strict_version_check=False) # start h2o

Checking whether there is an H2O instance running at http://localhost:54321 . connected.


--------------------------  ---------------------------------------------------
H2O cluster uptime:         2 mins 31 secs
H2O cluster timezone:       America/New_York
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.1
H2O cluster version age:    6 days
H2O cluster name:           H2O_from_python_newzysharma_p0cdw3
H2O cluster total nodes:    1
H2O cluster free memory:    2 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         locked, healthy
H2O connection url:         http://localhost:54321
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, XGBoost, Algos, AutoML, Core V3, Core V4
Python version:             3.7.1 final
--------------------------  ---------------------------------------------------

In [12]:
#importing data to the server
hp = h2o.import_file(path="hour.csv")

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [13]:
#Displaying the head
hp.head()

instant dteday season yr mnth hr holiday weekday workingday weathersit temp atemp hum windspeed casual registered cnt 
 
 
 1 2011-01-01 00:00:00 1 0 1 0 0 6 0 1 0.24 0.2879 0.81 0 3 13 16 
 2 2011-01-01 00:00:00 1 0 1 1 0 6 0 1 0.22 0.2727 0.8 0 8 32 40 
 3 2011-01-01 00:00:00 1 0 1 2 0 6 0 1 0.22 0.2727 0.8 0 5 27 32 
 4 2011-01-01 00:00:00 1 0 1 3 0 6 0 1 0.24 0.2879 0.75 0 3 10 13 
 5 2011-01-01 00:00:00 1 0 1 4 0 6 0 1 0.24 0.2879 0.75 0 0 1 1 
 6 2011-01-01 00:00:00 1 0 1 5 0 6 0 2 0.24 0.2576 0.75 0.0896 0 1 1 
 7 2011-01-01 00:00:00 1 0 1 6 0 6 0 1 0.22 0.2727 0.8 0 2 0 2 
 8 2011-01-01 00:00:00 1 0 1 7 0 6 0 1 0.2 0.2576 0.86 0 1 2 3 
 9 2011-01-01 00:00:00 1 0 1 8 0 6 0 1 0.24 0.2879 0.75 0 1 7 8 
 10 2011-01-01 00:00:00 1 0 1 9 0 6 0 1 0.32 0.3485 0.76 0 8 6 14

In [14]:
hp.describe()

Rows:17379
Cols:17




instant dteday season yr mnth hr holiday weekday workingday weathersit temp atemp hum windspeed casual registered cnt 
 
 
 type int time int int int int int int int int real real real real int int int 
 mins 1.0 1293840000000.0 1.0 0.0 1.0 0.0 0.0 0.0 0.0 1.0 0.02 0.0 0.0 0.0 0.0 0.0 1.0 
 mean 8690.0 1325477314552.0461 2.501639910236492 0.5025605615973301 6.5377754761493785 11.546751826917548 0.028770355026181024 3.003682605443351 0.6827205247712756 1.425283387997008 0.4969871684216584 0.47577510213476026 0.6272288394038784 0.1900976063064618 35.67621842453536 153.78686920996606 189.4630876345014 
 maxs 17379.0 1356912000000.0 4.0 1.0 12.0 23.0 1.0 6.0 1.0 4.0 1.0 1.0 1.0 0.8507 367.0 886.0 977.0 
 sigma 5017.029499614288 18150225217.779854 1.1069181394480765 0.5000078290910197 3.438775713750168 6.914405095264493 0.16716527638437123 2.005771456110988 0.4654306335238829 0.6393568777542534 0.19255612124972193 0.17185021563535932 0.19292983406291514 0.1223402285727905 49.30503038705309 151.35728591258314 181.38759909186476 
 zeros 0 0 0 8645 0 726 16879 2502 5514 0 0 2 22 2180 1581 24 0 
 missing 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 
 0 1.0 2011-01-01 00:00:00 1.0 0.0 1.0 0.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.81 0.0 3.0 13.0 16.0 
 1 2.0 2011-01-01 00:00:00 1.0 0.0 1.0 1.0 0.0 6.0 0.0 1.0 0.22 0.2727 0.8 0.0 8.0 32.0 40.0 
 2 3.0 2011-01-01 00:00:00 1.0 0.0 1.0 2.0 0.0 6.0 0.0 1.0 0.22 0.2727 0.8 0.0 5.0 27.0 32.0 
 3 4.0 2011-01-01 00:00:00 1.0 0.0 1.0 3.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.75 0.0 3.0 10.0 13.0 
 4 5.0 2011-01-01 00:00:00 1.0 0.0 1.0 4.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.75 0.0 0.0 1.0 1.0 
 5 6.0 2011-01-01 00:00:00 1.0 0.0 1.0 5.0 0.0 6.0 0.0 2.0 0.24 0.2576 0.75 0.0896 0.0 1.0 1.0 
 6 7.0 2011-01-01 00:00:00 1.0 0.0 1.0 6.0 0.0 6.0 0.0 1.0 0.22 0.2727 0.8 0.0 2.0 0.0 2.0 
 7 8.0 2011-01-01 00:00:00 1.0 0.0 1.0 7.0 0.0 6.0 0.0 1.0 0.2 0.2576 0.86 0.0 1.0 2.0 3.0 
 8 9.0 2011-01-01 00:00:00 1.0 0.0 1.0 8.0 0.0 6.0 0.0 1.0 0.24 0.2879 0.75 0.0 1.0 7.0 8.0 
 9 10.0 2011-01-01 00:00:00 1.0 0.0 1.0 9.0 0.0 6.0 0.0 1.0 0.32 0.3485 0.76 0.0 8.0 6.0 14.0

# Model with 500 seconds

In [37]:
# Assume the following are passed by the user from the web interface

'''
Need a user id and project id?

'''
target='cnt' 
data_file='hour.csv'
run_time=500
run_id='SOME_ID_20180617_221528' # Just some arbitrary ID
server_path='Users/newzysharma/Desktop/Desktop/Machine_Learning/Project'
classification=True
scale=False
max_models=None
balance_y=False # balance_classes=balance_y
balance_threshold=0.2
project ="HyperparameterDB_Project"  # project_name = project

In [42]:
# assign target and inputs for logistic regression
y = target
X = [name for name in df.columns if name != y]
print(y)
print(X)

cnt
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered']


In [43]:
# determine column types
ints, reals, enums = [], [], []
for key, val in df.types.items():
    if key in X:
        if val == 'enum':
            enums.append(key)
        elif val == 'int':
            ints.append(key)            
        else: 
            reals.append(key)

print(ints)
print(enums)
print(reals)

['instant', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'casual', 'registered']
[]
['dteday', 'temp', 'atemp', 'hum', 'windspeed']


In [44]:
# impute missing values
_ = df[reals].impute(method='mean')
_ = df[ints].impute(method='median')

if scale:
    df[reals] = df[reals].scale()
    df[ints] = df[ints].scale()

In [45]:
# set target to factor for classification by default or if user specifies classification
if classification:
    df[y] = df[y].asfactor()

In [46]:
df[y].levels()

[['1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '10',
  '11',
  '12',
  '13',
  '14',
  '15',
  '16',
  '17',
  '18',
  '19',
  '20',
  '21',
  '22',
  '23',
  '24',
  '25',
  '26',
  '27',
  '28',
  '29',
  '30',
  '31',
  '32',
  '33',
  '34',
  '35',
  '36',
  '37',
  '38',
  '39',
  '40',
  '41',
  '42',
  '43',
  '44',
  '45',
  '46',
  '47',
  '48',
  '49',
  '50',
  '51',
  '52',
  '53',
  '54',
  '55',
  '56',
  '57',
  '58',
  '59',
  '60',
  '61',
  '62',
  '63',
  '64',
  '65',
  '66',
  '67',
  '68',
  '69',
  '70',
  '71',
  '72',
  '73',
  '74',
  '75',
  '76',
  '77',
  '78',
  '79',
  '80',
  '81',
  '82',
  '83',
  '84',
  '85',
  '86',
  '87',
  '88',
  '89',
  '90',
  '91',
  '92',
  '93',
  '94',
  '95',
  '96',
  '97',
  '98',
  '99',
  '100',
  '101',
  '102',
  '103',
  '104',
  '105',
  '106',
  '107',
  '108',
  '109',
  '110',
  '111',
  '112',
  '113',
  '114',
  '115',
  '116',
  '117',
  '118',
  '119',
  '120',
  '121',
  '122',
  '123',
 


## Cross-validate rather than take a test training split with 500 seconds

In [47]:
# automl
# runs for run_time seconds then builds a stacked ensemble

aml = H2OAutoML(max_runtime_secs=run_time,project_name = project) # init automl, run for 500 seconds
aml.train(x=X,  
           y=y,
           training_frame=hp)

AutoML progress: |████████████████████████████████████████████████████████| 100%


## Leaderboard

In [48]:
# view leaderboard
lb = aml.leaderboard
lb

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 GBM_3_AutoML_20190407_222223 10.3938 3.22394 10.3938 1.79838 0.0255248 
 GBM_1_AutoML_20190407_222223 14.531 3.81196 14.531 2.39932 0.0535151 
 DeepLearning_1_AutoML_20190407_222223 14.694 3.83328 14.694 2.77942 nan 
 XGBoost_3_AutoML_20190407_222223 16.5115 4.06343 16.5115 2.65759 0.0686023 
 GBM_2_AutoML_20190407_222223 17.8634 4.22651 17.8634 2.61351 0.0574055 
 XGBoost_1_AutoML_20190408_000952 19.2132 4.38329 19.2132 2.5899 0.0497179 
 StackedEnsemble_AllModels_AutoML_20190407_222223 21.623 4.65005 21.623 3.14915 0.18555 
 XGBoost_3_AutoML_20190408_001901 25.3854 5.03839 25.3854 3.30893 0.09664 
 XGBoost_2_AutoML_20190408_000952 25.8932 5.08854 25.8932 2.88628 0.0524873 
 XGBoost_1_AutoML_20190407_222223 28.4128 5.33037 28.4128 3.35929 0.0995271

In [49]:
aml.leader

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_3_AutoML_20190407_222223


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 2.155227478290276
RMSE: 1.4680693029589154
MAE: 0.9893247655273629
RMSLE: 0.01894810024709767
Mean Residual Deviance: 2.155227478290276

ModelMetricsRegression: gbm
** Reported on cross-validation data. **

MSE: 10.39381683106471
RMSE: 3.2239442971404935
MAE: 1.798379228502425
RMSLE: 0.025524841564696307
Mean Residual Deviance: 10.39381683106471
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------
mae                     1.79838    0.0308816    1.83044       1.77347       1.73178       1.85713       1.79907
mean_residual_deviance  10.3938    0.578951     9.86147       9.70831       10.2984       11.9797       10.1212
mse                     10.3938    0.578951     9.86147       9.70831       10.2984       11.9797       10.1212
r2                      0.999684   1.76348e-05  0.999701      0.999707      0.999685      0.999637      0.999691
residual_deviance       10.3938    0.578951     9.86147       9.70831       10.2984       11.9797       10.1212
rmse                    3.22155    0.0877322    3.1403        3.11582       3.20911       3.46117       3.18138
rmsle                   0.0254593  0.00129373   0.0273739     0.0244253     0.0229992     0.0247157     0.0277825

Scoring History: 


timestamp            duration    number_of_trees    training_rmse       training_mae        training_deviance
---  -------------------  ----------  -----------------  ------------------  ------------------  -------------------
     2019-04-07 22:38:06  13.517 sec  0.0                181.38238043116908  142.3998489274107   32899.56793087735
     2019-04-07 22:38:06  13.582 sec  5.0                107.62882133004233  84.42612339315882   11583.963180894176
     2019-04-07 22:38:06  13.653 sec  10.0               63.87971533241224   50.02755951449267   4080.6180309500232
     2019-04-07 22:38:06  13.722 sec  15.0               38.308109257036094  29.932374587767733  1467.5112348490143
     2019-04-07 22:38:06  13.791 sec  20.0               22.832737245523866  17.78686751399629   521.3338901231327
---  ---                  ---         ---                ---                 ---                 ---
     2019-04-07 22:38:08  15.781 sec  185.0              1.5150515480600888  1.0157185701487343  2.2953811932792716
     2019-04-07 22:38:08  15.834 sec  190.0              1.50039536201263    1.0083876913453598  2.251186242349011
     2019-04-07 22:38:08  15.886 sec  195.0              1.4894063572137368  1.0026039381324645  2.2183312969086937
     2019-04-07 22:38:08  15.938 sec  200.0              1.4792239511275211  0.995939672974148   2.188103497589315
     2019-04-07 22:38:08  15.979 sec  204.0              1.4680693029589154  0.9893247655273629  2.155227478290276


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
registered  2.52218e+09            1                    0.901929
casual      1.73936e+08            0.0689626            0.0621994
hr          5.33077e+07            0.0211356            0.0190628
workingday  2.27517e+07            0.00902062           0.00813596
temp        1.00548e+07            0.00398655           0.00359558
dteday      6.58149e+06            0.00260944           0.00235353
atemp       1.72435e+06            0.000683674          0.000616625
weekday     1.57954e+06            0.000626261          0.000564842
weathersit  1.21686e+06            0.000482465          0.000435149
instant     1.13507e+06            0.000450034          0.000405898
mnth        779646                 0.000309116          0.0002788
yr          685794                 0.000271905          0.000245239
hum         294052                 0.000116586          0.000105153
season      116064                 4.60173e-05          4.15043e-05
windspeed   67994.2                2.69585e-05          2.43146e-05
holiday     18668.3                7.40163e-06          6.67574e-06

In [56]:
aml.leader.algo

'gbm'

In [57]:
dir(aml.leader)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_bc',
 '_bcin',
 '_check_and_save_parm',
 '_check_targets',
 '_compute_algo',
 '_end_time',
 '_estimator_type',
 '_future',
 '_get_metrics',
 '_have_mojo',
 '_have_pojo',
 '_id',
 '_is_xvalidated',
 '_job',
 '_keyify_if_h2oframe',
 '_make_model',
 '_metrics_class',
 '_model_json',
 '_parms',
 '_plot',
 '_requires_training_frame',
 '_resolve_model',
 '_run_time',
 '_start_time',
 '_train',
 '_verify_training_frame_params',
 '_xval_keys',
 'actual_params',
 'aic',
 'algo',
 'auc',
 'balance_classes',
 'biases',
 'build_tree_one_node',
 'calibrate_model',
 'calibration_frame',
 'categorical_encoding',
 'cato

## Ensemble Exploration

In [50]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
aml_leaderboard_df

model_id  mean_residual_deviance  \
0                        GBM_3_AutoML_20190407_222223               10.393817   
1                        GBM_1_AutoML_20190407_222223               14.531027   
2               DeepLearning_1_AutoML_20190407_222223               14.694030   
3                    XGBoost_3_AutoML_20190407_222223               16.511499   
4                        GBM_2_AutoML_20190407_222223               17.863356   
5                    XGBoost_1_AutoML_20190408_000952               19.213204   
6    StackedEnsemble_AllModels_AutoML_20190407_222223               21.622982   
7                    XGBoost_3_AutoML_20190408_001901               25.385364   
8                    XGBoost_2_AutoML_20190408_000952               25.893189   
9                    XGBoost_1_AutoML_20190407_222223               28.412796   
10                   XGBoost_1_AutoML_20190408_001901               29.321439   
11                   XGBoost_2_AutoML_20190407_222223               29.931125   
12                   XGBoost_2_AutoML_20190408_001901               30.280936   
13                       GBM_4_AutoML_20190407_222223               33.269049   
14          GLM_grid_1_AutoML_20190408_001901_model_1               38.387049   
15          GLM_grid_1_AutoML_20190407_222223_model_1               38.387049   
16          GLM_grid_1_AutoML_20190408_000952_model_1               38.387049   
17                       GBM_5_AutoML_20190407_222223               40.097180   
18                       DRF_1_AutoML_20190407_222223               46.224341   
19  StackedEnsemble_BestOfFamily_AutoML_20190407_2...               55.264544   
20                   XGBoost_3_AutoML_20190408_000952               58.236630   
21                       XRT_1_AutoML_20190407_222223               75.849977   
22                       DRF_1_AutoML_20190408_001901               86.047669   
23                       DRF_1_AutoML_20190408_000952              100.191599   

         rmse         mse       mae     rmsle  
0    3.223944   10.393817  1.798379  0.025525  
1    3.811958   14.531027  2.399321  0.053515  
2    3.833279   14.694030  2.779418       NaN  
3    4.063434   16.511499  2.657589  0.068602  
4    4.226506   17.863356  2.613506  0.057406  
5    4.383287   19.213204  2.589897  0.049718  
6    4.650052   21.622982  3.149146  0.185550  
7    5.038389   25.385364  3.308930  0.096640  
8    5.088535   25.893189  2.886280  0.052487  
9    5.330365   28.412796  3.359290  0.099527  
10   5.414927   29.321439  3.266991  0.064156  
11   5.470935   29.931125  3.504274       NaN  
12   5.502812   30.280936  3.057353  0.070078  
13   5.767933   33.269049  3.793576       NaN  
14   6.195728   38.387049  4.451969       NaN  
15   6.195728   38.387049  4.451969       NaN  
16   6.195728   38.387049  4.451969       NaN  
17   6.332233   40.097180  3.418618  0.064996  
18   6.798848   46.224341  3.722796  0.078268  
19   7.434013   55.264544  5.389613  0.298504  
20   7.631293   58.236630  5.083642       NaN  
21   8.709189   75.849977  5.072365  0.096378  
22   9.276188   86.047669  5.338030  0.096408  
23  10.009575  100.191599  5.757416  0.101491

## Getting Models

In [53]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[1])

In [52]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'XGBoost_1_AutoML_20190408_000952',
   'type': 'Key<Model>',
   'URL': '/3/Models/XGBoost_1_AutoML_20190408_000952'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour1.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour1.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'fold_assignment': {'default': 'AUTO', 'actual': 'Modulo'},
 'fold_column': {'def

# Model with runtime 1000 seconds

In [10]:
# Assume the following are passed by the user from the web interface

'''
Need a user id and project id?

'''
target='cnt' 
data_file='hour.csv'
run_time=1000
run_id='SOME_ID_20180617_221529' # Just some arbitrary ID
server_path='Users/newzysharma/Desktop/Desktop/Machine_Learning/Project'
classification=True
scale=False
max_models=None
balance_y=False # balance_classes=balance_y
balance_threshold=0.2
project ="HyperparameterDB_Project"  # project_name = project

In [22]:
# Use local data file or download from some type of bucket
import os

data_path=os.path.join(server_path,data_file)
data_path

'Users/newzysharma/Desktop/Desktop/Machine_Learning/Project/hour.csv'

In [11]:
# assign target and inputs for logistic regression
y = target
X = [name for name in df.columns if name != y]
print(y)
print(X)

cnt
['instant', 'dteday', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'temp', 'atemp', 'hum', 'windspeed', 'casual', 'registered']


In [18]:
# determine column types
ints, reals, enums = [], [], []
for key, val in df.types.items():
    if key in X:
        if val == 'enum':
            enums.append(key)
        elif val == 'int':
            ints.append(key)            
        else: 
            reals.append(key)

print(ints)
print(enums)
print(reals)

['instant', 'season', 'yr', 'mnth', 'hr', 'holiday', 'weekday', 'workingday', 'weathersit', 'casual', 'registered']
[]
['dteday', 'temp', 'atemp', 'hum', 'windspeed']


In [19]:
# impute missing values
_ = df[reals].impute(method='mean')
_ = df[ints].impute(method='median')

if scale:
    df[reals] = df[reals].scale()
    df[ints] = df[ints].scale()

In [20]:
# set target to factor for classification by default or if user specifies classification
if classification:
    df[y] = df[y].asfactor()

In [21]:
df[y].levels()

[['1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '10',
  '11',
  '12',
  '13',
  '14',
  '15',
  '16',
  '17',
  '18',
  '19',
  '20',
  '21',
  '22',
  '23',
  '24',
  '25',
  '26',
  '27',
  '28',
  '29',
  '30',
  '31',
  '32',
  '33',
  '34',
  '35',
  '36',
  '37',
  '38',
  '39',
  '40',
  '41',
  '42',
  '43',
  '44',
  '45',
  '46',
  '47',
  '48',
  '49',
  '50',
  '51',
  '52',
  '53',
  '54',
  '55',
  '56',
  '57',
  '58',
  '59',
  '60',
  '61',
  '62',
  '63',
  '64',
  '65',
  '66',
  '67',
  '68',
  '69',
  '70',
  '71',
  '72',
  '73',
  '74',
  '75',
  '76',
  '77',
  '78',
  '79',
  '80',
  '81',
  '82',
  '83',
  '84',
  '85',
  '86',
  '87',
  '88',
  '89',
  '90',
  '91',
  '92',
  '93',
  '94',
  '95',
  '96',
  '97',
  '98',
  '99',
  '100',
  '101',
  '102',
  '103',
  '104',
  '105',
  '106',
  '107',
  '108',
  '109',
  '110',
  '111',
  '112',
  '113',
  '114',
  '115',
  '116',
  '117',
  '118',
  '119',
  '120',
  '121',
  '122',
  '123',
 

In [23]:
if classification:
    class_percentage = y_balance=df[y].mean()[0]/(df[y].max()-df[y].min())
    if class_percentage < balance_threshold:
        balance_y=True
        

print(run_time)
type(run_time)

1000


int


## Cross-validate rather than take a test training split with 1000 seconds

In [24]:
# automl
# runs for run_time seconds then builds a stacked ensemble

aml = H2OAutoML(max_runtime_secs=run_time,project_name = project) # init automl, run for 1000 seconds
aml.train(x=X,  
           y=y,
           training_frame=hp)

AutoML progress: |████████████████████████████████████████████████████████| 100%


## Leaderboard

In [25]:
# view leaderboard
lb = aml.leaderboard
lb

model_id mean_residual_deviance rmse mse mae rmsle 
 
 
 GBM_3_AutoML_20190407_222223 10.3938 3.22394 10.3938 1.79838 0.0255248 
 GBM_1_AutoML_20190407_222223 14.531 3.81196 14.531 2.39932 0.0535151 
 DeepLearning_1_AutoML_20190407_222223 14.694 3.83328 14.694 2.77942 nan 
 XGBoost_3_AutoML_20190407_222223 16.5115 4.06343 16.5115 2.65759 0.0686023 
 GBM_2_AutoML_20190407_222223 17.8634 4.22651 17.8634 2.61351 0.0574055 
 StackedEnsemble_AllModels_AutoML_20190407_222223 21.623 4.65005 21.623 3.14915 0.18555 
 XGBoost_1_AutoML_20190407_222223 28.4128 5.33037 28.4128 3.35929 0.0995271 
 XGBoost_2_AutoML_20190407_222223 29.9311 5.47093 29.9311 3.50427 nan 
 GBM_4_AutoML_20190407_222223 33.269 5.76793 33.269 3.79358 nan 
 GLM_grid_1_AutoML_20190407_222223_model_1 38.387 6.19573 38.387 4.45197 nan

In [26]:
aml.leader

Model Details
H2OGradientBoostingEstimator :  Gradient Boosting Machine
Model Key:  GBM_3_AutoML_20190407_222223


ModelMetricsRegression: gbm
** Reported on train data. **

MSE: 2.155227478290276
RMSE: 1.4680693029589154
MAE: 0.9893247655273629
RMSLE: 0.01894810024709767
Mean Residual Deviance: 2.155227478290276

ModelMetricsRegression: gbm
** Reported on cross-validation data. **

MSE: 10.39381683106471
RMSE: 3.2239442971404935
MAE: 1.798379228502425
RMSLE: 0.025524841564696307
Mean Residual Deviance: 10.39381683106471
Cross-Validation Metrics Summary: 


mean       sd           cv_1_valid    cv_2_valid    cv_3_valid    cv_4_valid    cv_5_valid
----------------------  ---------  -----------  ------------  ------------  ------------  ------------  ------------
mae                     1.79838    0.0308816    1.83044       1.77347       1.73178       1.85713       1.79907
mean_residual_deviance  10.3938    0.578951     9.86147       9.70831       10.2984       11.9797       10.1212
mse                     10.3938    0.578951     9.86147       9.70831       10.2984       11.9797       10.1212
r2                      0.999684   1.76348e-05  0.999701      0.999707      0.999685      0.999637      0.999691
residual_deviance       10.3938    0.578951     9.86147       9.70831       10.2984       11.9797       10.1212
rmse                    3.22155    0.0877322    3.1403        3.11582       3.20911       3.46117       3.18138
rmsle                   0.0254593  0.00129373   0.0273739     0.0244253     0.0229992     0.0247157     0.0277825

Scoring History: 


timestamp            duration    number_of_trees    training_rmse       training_mae        training_deviance
---  -------------------  ----------  -----------------  ------------------  ------------------  -------------------
     2019-04-07 22:38:06  13.517 sec  0.0                181.38238043116908  142.3998489274107   32899.56793087735
     2019-04-07 22:38:06  13.582 sec  5.0                107.62882133004233  84.42612339315882   11583.963180894176
     2019-04-07 22:38:06  13.653 sec  10.0               63.87971533241224   50.02755951449267   4080.6180309500232
     2019-04-07 22:38:06  13.722 sec  15.0               38.308109257036094  29.932374587767733  1467.5112348490143
     2019-04-07 22:38:06  13.791 sec  20.0               22.832737245523866  17.78686751399629   521.3338901231327
---  ---                  ---         ---                ---                 ---                 ---
     2019-04-07 22:38:08  15.781 sec  185.0              1.5150515480600888  1.0157185701487343  2.2953811932792716
     2019-04-07 22:38:08  15.834 sec  190.0              1.50039536201263    1.0083876913453598  2.251186242349011
     2019-04-07 22:38:08  15.886 sec  195.0              1.4894063572137368  1.0026039381324645  2.2183312969086937
     2019-04-07 22:38:08  15.938 sec  200.0              1.4792239511275211  0.995939672974148   2.188103497589315
     2019-04-07 22:38:08  15.979 sec  204.0              1.4680693029589154  0.9893247655273629  2.155227478290276


See the whole table with table.as_data_frame()
Variable Importances: 


variable    relative_importance    scaled_importance    percentage
----------  ---------------------  -------------------  ------------
registered  2.52218e+09            1                    0.901929
casual      1.73936e+08            0.0689626            0.0621994
hr          5.33077e+07            0.0211356            0.0190628
workingday  2.27517e+07            0.00902062           0.00813596
temp        1.00548e+07            0.00398655           0.00359558
dteday      6.58149e+06            0.00260944           0.00235353
atemp       1.72435e+06            0.000683674          0.000616625
weekday     1.57954e+06            0.000626261          0.000564842
weathersit  1.21686e+06            0.000482465          0.000435149
instant     1.13507e+06            0.000450034          0.000405898
mnth        779646                 0.000309116          0.0002788
yr          685794                 0.000271905          0.000245239
hum         294052                 0.000116586          0.000105153
season      116064                 4.60173e-05          4.15043e-05
windspeed   67994.2                2.69585e-05          2.43146e-05
holiday     18668.3                7.40163e-06          6.67574e-06

In [58]:
aml.leader.algo

'gbm'

In [59]:
dir(aml.leader)

['__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattr__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 '_bc',
 '_bcin',
 '_check_and_save_parm',
 '_check_targets',
 '_compute_algo',
 '_end_time',
 '_estimator_type',
 '_future',
 '_get_metrics',
 '_have_mojo',
 '_have_pojo',
 '_id',
 '_is_xvalidated',
 '_job',
 '_keyify_if_h2oframe',
 '_make_model',
 '_metrics_class',
 '_model_json',
 '_parms',
 '_plot',
 '_requires_training_frame',
 '_resolve_model',
 '_run_time',
 '_start_time',
 '_train',
 '_verify_training_frame_params',
 '_xval_keys',
 'actual_params',
 'aic',
 'algo',
 'auc',
 'balance_classes',
 'biases',
 'build_tree_one_node',
 'calibrate_model',
 'calibration_frame',
 'categorical_encoding',
 'cato

## Ensemble Exploration

In [27]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
aml_leaderboard_df

model_id  mean_residual_deviance  \
0                        GBM_3_AutoML_20190407_222223               10.393817   
1                        GBM_1_AutoML_20190407_222223               14.531027   
2               DeepLearning_1_AutoML_20190407_222223               14.694030   
3                    XGBoost_3_AutoML_20190407_222223               16.511499   
4                        GBM_2_AutoML_20190407_222223               17.863356   
5    StackedEnsemble_AllModels_AutoML_20190407_222223               21.622982   
6                    XGBoost_1_AutoML_20190407_222223               28.412796   
7                    XGBoost_2_AutoML_20190407_222223               29.931125   
8                        GBM_4_AutoML_20190407_222223               33.269049   
9           GLM_grid_1_AutoML_20190407_222223_model_1               38.387049   
10                       GBM_5_AutoML_20190407_222223               40.097180   
11                       DRF_1_AutoML_20190407_222223               46.224341   
12  StackedEnsemble_BestOfFamily_AutoML_20190407_2...               55.264544   
13                       XRT_1_AutoML_20190407_222223               75.849977   

        rmse        mse       mae     rmsle  
0   3.223944  10.393817  1.798379  0.025525  
1   3.811958  14.531027  2.399321  0.053515  
2   3.833279  14.694030  2.779418       NaN  
3   4.063434  16.511499  2.657589  0.068602  
4   4.226506  17.863356  2.613506  0.057406  
5   4.650052  21.622982  3.149146  0.185550  
6   5.330365  28.412796  3.359290  0.099527  
7   5.470935  29.931125  3.504274       NaN  
8   5.767933  33.269049  3.793576       NaN  
9   6.195728  38.387049  4.451969       NaN  
10  6.332233  40.097180  3.418618  0.064996  
11  6.798848  46.224341  3.722796  0.078268  
12  7.434013  55.264544  5.389613  0.298504  
13  8.709189  75.849977  5.072365  0.096378

##### As per the rmse metric that "The smaller the RMSE value, the better the model".
##### As per the mse metric that "The smaller the MSE value, the better the model".
So, In our case GBM_3_AutoML_20190407_222223 has the smallest rmse = 3.223944 and mse = 10.393817 that is why it is our best model.

In [33]:
aml_leaderboard_df.head()

model_id  mean_residual_deviance      rmse  \
0           GBM_3_AutoML_20190407_222223               10.393817  3.223944   
1           GBM_1_AutoML_20190407_222223               14.531027  3.811958   
2  DeepLearning_1_AutoML_20190407_222223               14.694030  3.833279   
3       XGBoost_3_AutoML_20190407_222223               16.511499  4.063434   
4           GBM_2_AutoML_20190407_222223               17.863356  4.226506   

         mse       mae     rmsle  
0  10.393817  1.798379  0.025525  
1  14.531027  2.399321  0.053515  
2  14.694030  2.779418       NaN  
3  16.511499  2.657589  0.068602  
4  17.863356  2.613506  0.057406

## Getting Models

In [35]:
aml_leaderboard_df=aml.leaderboard.as_data_frame()
model_set=aml_leaderboard_df['model_id']
mod_best=h2o.get_model(model_set[5])

In [31]:
mod_best.params

{'model_id': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'ModelKeyV3',
    'schema_type': 'Key<Model>'},
   'name': 'GBM_1_AutoML_20190407_222223',
   'type': 'Key<Model>',
   'URL': '/3/Models/GBM_1_AutoML_20190407_222223'}},
 'training_frame': {'default': None,
  'actual': {'__meta': {'schema_version': 3,
    'schema_name': 'FrameKeyV3',
    'schema_type': 'Key<Frame>'},
   'name': 'automl_training_hour1.hex',
   'type': 'Key<Frame>',
   'URL': '/3/Frames/automl_training_hour1.hex'}},
 'validation_frame': {'default': None, 'actual': None},
 'nfolds': {'default': 0, 'actual': 5},
 'keep_cross_validation_models': {'default': True, 'actual': False},
 'keep_cross_validation_predictions': {'default': False, 'actual': True},
 'keep_cross_validation_fold_assignment': {'default': False, 'actual': False},
 'score_each_iteration': {'default': False, 'actual': False},
 'score_tree_interval': {'default': 0, 'actual': 5},
 'fold_assignment': {'default': 'AUTO